### PR TITLE
Add API documentation to generated C files.

### PR DESCRIPTION
**Pull request by duesee**
This PR adds documentation to the generated C files. We created an ad-hoc documentation for various algorithms for [HACL Packages](https://tech.cryspen.com/hacl-packages/c/main/hacl/aead/chacha20poly1305.html) and now want to transfer the documentation where it belongs.

### CODE REVIEW

**Review by R1kM — APPROVED** [review]
Thanks! I left some minor comments, but this looks good to me overall

File: code/blake2/Hacl.Blake2b_256.fst [hunk 1]
```diff
@@ -36,6 +36,14 @@ let blake2b_finish : Impl.blake2_finish_st Spec.Blake2B Core.M256 =
   Impl.blake2_finish #Spec.Blake2B #Core.M256
 
 (* The one-shot hash *)
+[@@ Comment "Write the BLAKE2b digest of message `d` using key `k` into `output`.
+
+@param nn Length of the to-be-generated digest with 1 < `nn` <= 64.
```
[the remark this answers is not in the document]
Comment (duesee) [reply code/blake2/Hacl.Blake2b_256.fst:41]: Good catch! No reason for `1 < nn`. That was an off-by-one mistake I made.

File: code/curve25519/Hacl.Curve25519_51.fsti [hunk 2]
```diff
@@ -3,6 +3,24 @@ module Hacl.Curve25519_51
 open Hacl.Impl.Curve25519.Generic
 open Hacl.Impl.Curve25519.Fields
 
+[@@ Comment "Compute the scalar multiple of a point.
```
Comment (polubelova) [inline code/curve25519/Hacl.Curve25519_51.fsti:6]: Should we add the length of public and secret keys?

File: code/curve25519/Hacl.Curve25519_64_Slow.fsti [hunk 3]
```diff
@@ -3,6 +3,24 @@ module Hacl.Curve25519_64_Slow
 open Hacl.Impl.Curve25519.Generic
 open Hacl.Impl.Curve25519.Fields
 
+[@@ Comment "Compute the scalar multiple of a point.
```
Comment (polubelova) [inline code/curve25519/Hacl.Curve25519_64_Slow.fsti:6]: I think we need to remove this file, as it's not supposed to be used. I can do it later

File: code/curve25519/Hacl.Curve25519_64.fsti [hunk 4]
```diff
@@ -6,6 +6,24 @@ open Hacl.Impl.Curve25519.Fields
 inline_for_extraction noextract
 let p = Vale.X64.CPU_Features_s.(adx_enabled /\ bmi2_enabled)
 
+[@@ Comment "Compute the scalar multiple of a point.
```
[the remark this answers is not in the document]
Comment (duesee) [reply code/curve25519/Hacl.Curve25519_64.fsti:9]: I'm not sure... Having this in the doc comment could be helpful. However, I am worried a bit that the docs may become too verbose. From my understanding these requirements mostly apply to "full algorithms" and not only functions. Thus, in the HACL Packages documentation, we introduce the requirements upfront and do not mention them later. An early version of that can be seen in the [Curve25519 section](https://tech.cryspen.com/hacl-packages/c/main/hacl/dh/ec/curve25519.html).

Do you think this is sufficient?

[the remark this answers is not in the document]
Comment (R1kM) [reply code/curve25519/Hacl.Curve25519_64.fsti:9]: Indeed, I think having them at the top of the header would work best.
I'm fine with the current version

[the remark this answers is not in the document]
Comment (duesee) [reply code/curve25519/Hacl.Curve25519_64.fsti:9]: Cool, so let's get this in. We will probably make a bunch of other PRs in the future and can take care of that later.

File: code/blake2/Hacl.Blake2b_256.fst (hunk 1 above)
[the remark this answers is not in the document]
Comment (duesee) [reply code/blake2/Hacl.Blake2b_256.fst]: Fixed-up in [98da5ec](https://github.com/hacl-star/hacl-star/pull/667/commits/98da5ec7f8a1d9c74dec1e8728bd2eaa3d4afbee).

File: code/nacl-box/Hacl.NaCl.fst [hunk 5]
```diff
@@ -16,6 +16,14 @@ module SS = Spec.SecretBox
 
 #set-options "--max_fuel 50 --max_fuel 0 --max_ifuel 0"
 
+[@@ Comment "Encrypt a message with a key and nonce.
```
Comment (polubelova) [inline code/nacl-box/Hacl.NaCl.fst:19]: Should we add a comment saying that encryption and decryption can be done in-place?

**Comment by duesee** [comment]
Ah, I see you already approved the changes. Let me just rebase, fixup, and force-push the commits. This should be good to go then.

File: code/nacl-box/Hacl.NaCl.fst [hunk 6]
```diff
@@ -99,6 +129,11 @@ let crypto_secretbox_open_easy m c clen n k =
   Hacl.Impl.SecretBox.secretbox_open_easy (clen -! 16ul) m k n c
 
 
+[@@ Comment "Compute a shared secret key given a public key and secret key.
```
Comment (polubelova) [inline code/nacl-box/Hacl.NaCl.fst:132]: Should we specify the length of `pk` and `sk`?

**Comment by duesee** [comment]
Ready :-)

**Review by polubelova — APPROVED** [review]
Looks good to me, thanks!

File: code/nacl-box/Hacl.NaCl.fst (hunk 5 above)
Comment (duesee) [reply code/nacl-box/Hacl.NaCl.fst:19]: Good point. I'll add something :+1:

File: code/curve25519/Hacl.Curve25519_64_Slow.fsti (hunk 3 above)
Comment (duesee) [reply code/curve25519/Hacl.Curve25519_64_Slow.fsti:6]: Sounds good, thanks! I'll leave this part of the PR unchanged.

File: code/curve25519/Hacl.Curve25519_51.fsti (hunk 2 above)
Comment (duesee) [reply code/curve25519/Hacl.Curve25519_51.fsti:6]: Added in 2cbd6d0f35d3859338fd736d64c3bc8c8ba47a62.

File: code/nacl-box/Hacl.NaCl.fst (hunk 5 above)
Comment (duesee) [reply code/nacl-box/Hacl.NaCl.fst:19]: Added in 239dc5c7ef2d81f9a47d0d296ca41b2b22c16404.

File: code/nacl-box/Hacl.NaCl.fst [hunk 7]
```diff
@@ -180,6 +230,15 @@ let crypto_box_open_detached_afternm m c tag mlen n k =
...
 
 
+[@@ Comment "Verify and decrypt a ciphertext produced by `crypto_box_detached`.
+
+@param m Pointer to `mlen` bytes of memory where the decrypted message is written to.
+@param c Pointer to `mlen` bytes of memory where the ciphertext is read from. Note: the ciphertext must include the tag.
+@param tag Pointer to 16 (tag length) bytes of memory where the authentication tag is read from.
+@param mlen Length of the message (and ciphertext).
+@param n Pointer to 24 (`crypto_box_NONCEBYTES`) bytes of memory where the nonce is read from.
+@param pk Pointer to public key of the sender.
```
Comment (duesee) [inline code/nacl-box/Hacl.NaCl.fst:240]: Add pk sk len here, too.

File: code/nacl-box/Hacl.NaCl.fst [hunk 8]
```diff
@@ -262,6 +331,14 @@ let crypto_box_open_easy_afternm m c clen n k =
...
 
 
+[@@ Comment "Verify and decrypt a ciphertext produced by `crypto_box_easy`.
+
+@param m Pointer to `clen` - 16 (tag length) bytes of memory where the decrypted message is written to.
+@param c Pointer to `clen` bytes of memory where the ciphertext is read from. Note: the ciphertext must include the tag.
+@param clen Length of the ciphertext.
+@param n Pointer to 24 (`crypto_box_NONCEBYTES`) bytes of memory where the nonce is read from.
+@param pk Pointer to public key of the sender.
+@param sk Pointer to secret key of the recipient."]
```
Comment (duesee) [inline code/nacl-box/Hacl.NaCl.fst:341]: Add pk sk len here, too.

File: code/nacl-box/Hacl.NaCl.fst (hunk 6 above)
Comment (duesee) [reply code/nacl-box/Hacl.NaCl.fst:136]: Added in later fixups.

File: code/nacl-box/Hacl.NaCl.fst (hunk 7 above)
Comment (duesee) [reply code/nacl-box/Hacl.NaCl.fst:240]: Added in later fixups.

File: code/nacl-box/Hacl.NaCl.fst (hunk 8 above)
Comment (duesee) [reply code/nacl-box/Hacl.NaCl.fst]: Added in later fixups.

**Comment by duesee** [comment]
Thanks for the review, @polubelova! I've added some more pk sk lengths.

File: code/curve25519/Hacl.Curve25519_64.fsti (hunk 4 above)
Comment (polubelova) [inline code/curve25519/Hacl.Curve25519_64.fsti:9]: Do you want to update comments here similar to `Hacl.Curve25519_51.fsti`?

**Comment by duesee** [comment]
I can clean up the commits after someone gives me a sign to do it to not invalidate the review notes until this can be merged.

**Comment by polubelova** [comment]
Thanks @duesee! Once you rebase your branch, we can merge it into `main`.

**Pull request merged**